pkg/steps/testing.go

```diff
@@ -44,6 +83,90 @@ func (dl *DryLogger) GetObjects() []runtime.Object {
 	return dl.objects
 }
 
+func (dl *DryLogger) AddBuild(build *buildapi.Build) {
+	imageLabels := build.Spec.Output.ImageLabels
+	if len(imageLabels) > 0 {
+		sort.Slice(imageLabels, func(i, j int) bool {
+			return imageLabels[i].Name < imageLabels[j].Name
+		})
+	}
+
+	if images := build.Spec.Source.Images; len(images) > 0 {
+		for i, image := range images {
+			build.Spec.Source.Images[i].From.Name = trimSHA256(image.From.Name)
+		}
+
+	}
+
+	dl.AddObject(build.DeepCopyObject())
+}
+
+func (dl *DryLogger) AddImageStreamTag(ist *imageapi.ImageStreamTag) {
+	if tag := ist.Tag; tag != nil {
+		if tag.From != nil {
+			ist.Tag.From.Name = trimSHA256(ist.Tag.From.Name)
+		}
+	}
+	dl.AddObject(ist.DeepCopyObject())
+}
+
+func trimSHA256(value string) string {
+	reg := regexp.MustCompile("@sha256:[0-9a-f]{64}")
+	return reg.ReplaceAllString(value, "@sha256:SHA")
+}
+
+// Log encode/decode the objects and print them as a valid JSON array.
+func (dl *DryLogger) Log() error {
+	var errors []error
+	objects := dl.objects
+	for _, object := range objects {
+		var b []byte
+		buffer := bytes.NewBuffer(b)
+
+		err := encoder.Encode(object, buffer)
+		if err != nil {
+			fmt.Printf("Unexpected encode error '%v'\n", err)
+			errors = append(errors, err)
+			continue
+		}
+
+		_, _, err = decoder.Decode(buffer.Bytes(), nil, object)
+		if err != nil {
+			fmt.Printf("Unexpected decode error %v\n", err)
+			errors = append(errors, err)
+			continue
+		}
+	}
+
+	sort.Slice(objects, func(i, j int) bool {
+		iAccessor, err := meta.Accessor(objects[i])
+		if err != nil {
+			errors = append(errors, fmt.Errorf("couldn't create accessor: %v", err))
+		}
+		jAccessor, err := meta.Accessor(objects[j])
+		if err != nil {
+			errors = append(errors, fmt.Errorf("couldn't create accessor: %v", err))
+		}
+
+		iKind := objects[i].GetObjectKind().GroupVersionKind().Kind
+		jKind := objects[j].GetObjectKind().GroupVersionKind().Kind
+
+		return iKind+iAccessor.GetName() < jKind+jAccessor.GetName()
```

Comment:
This looks strange, why not do two comparisons?

Reply:
There are objects with the same name. By adding `kind` we make sure that the ordering will be more stable.

Reply:
By "two comparisons", I meant first compare the kinds, and only if they are the same compare the names.